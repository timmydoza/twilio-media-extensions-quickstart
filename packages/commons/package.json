{
  "name": "@media-extensions/commons",
  "version": "1.0.0",
  "private": true,
  "main": "build/index.js",
  "types": "build/index.d.ts",
  "devDependencies": {
    "@testing-library/jest-dom": "^5.11.4",
    "@testing-library/react": "^11.1.0",
    "@testing-library/react-hooks": "^7.0.2",
    "@types/react": "^17.0.39",
    "@types/jest": "^26.0.15",
    "react": "^17.0.2",
    "ts-jest": "^26.5.6",
    "twilio-video": "^2.19.1",
    "typescript": "^4.1.2"
  },
  "scripts": {
    "clean": "rm -rf ./build && rm -rf tsconfig.tsbuildinfo",
    "compile": "tsc -b",
    "build": "npm run clean && npm run compile",
    "test": "jest"
  }
}
