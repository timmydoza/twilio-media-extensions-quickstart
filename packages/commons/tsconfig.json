{
  "compilerOptions": {
    "composite": true,
    "outDir": "build",
    "rootDir": "src",
    "module": "commonjs",
    "target": "es5",
    "declaration": true,
    "moduleResolution": "node",
    "allowSyntheticDefaultImports": true,
    "jsx": "react-jsx",
    "skipLibCheck": true,
    "esModuleInterop": true,
  },
  "include": ["src"],
}
