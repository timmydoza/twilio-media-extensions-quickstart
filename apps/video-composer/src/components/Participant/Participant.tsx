import { RemoteAudioTrack, RemoteParticipant, RemoteVideoTrack } from 'twilio-video';
import { useIsTrackEnabled, useTracks, useIsTrackSwitchedOff } from '@media-extensions/commons';
import { MARGIN } from '../../constants';
import MediaTrack from '../MediaTrack/MediaTrack';
import MutedIcon from './MutedIcon';

const borderWidth = 3;

export default function Participant({
  participant,
  width,
  height,
  isDominantSpeaker,
  fontSize,
}: {
  participant: RemoteParticipant;
  width: string;
  height: string;
  isDominantSpeaker: boolean;
  fontSize: string;
}) {
  const tracks = useTracks(participant);

  const audioTrack = tracks.find((track) => track.kind === 'audio') as RemoteAudioTrack;
  const isAudioEnabled = useIsTrackEnabled(audioTrack);

  const videoTrack = tracks.find(
    (track) => track.kind === 'video' && track.name !== 'video-composer-presentation'
  ) as RemoteVideoTrack;
  const isVideoEnabled = useIsTrackEnabled(videoTrack);
  const isVideoTrackSwitchedOff = useIsTrackSwitchedOff(videoTrack);

  return (
    <div
      className="participant"
      style={{
        width,
        height,
        margin: isDominantSpeaker ? MARGIN - borderWidth : MARGIN,
        border: isDominantSpeaker ? `solid ${borderWidth}px #7BEAA5` : 'none',
      }}
      data-cy-participant={participant.identity}
    >
      {!isAudioEnabled && <MutedIcon />}
      {audioTrack && <MediaTrack track={audioTrack} />}

      {(!isVideoEnabled || isVideoTrackSwitchedOff) && (
        <div style={{ fontSize }} className="noVideo">
          {!window.context?.ENV_hideParticipantIdentities && participant.identity}
        </div>
      )}

      {videoTrack && <MediaTrack track={videoTrack} />}

      {!window.context?.ENV_hideParticipantIdentities && <p>{participant.identity}</p>}
    </div>
  );
}
