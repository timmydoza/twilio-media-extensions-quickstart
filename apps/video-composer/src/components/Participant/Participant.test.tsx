import { shallow } from 'enzyme';
import { useTracks } from '@media-extensions/commons';
import MutedIcon from './MutedIcon';
import Participant from './Participant';

jest.mock('@media-extensions/commons', () => ({
  ...jest.requireActual('@media-extensions/commons'),
  useTracks: jest.fn(),
}));

const mockUseTracks = useTracks as jest.Mock<any>;

describe('the Participant component', () => {
  beforeEach(() => {
    mockUseTracks.mockImplementation(() => [
      { trackName: 'camera-123456', kind: 'video', trackSid: 'testVideoSid', isEnabled: true },
      { trackName: 'microphone-123456', kind: 'audio', trackSid: 'testAudioSid', isEnabled: true },
    ]);

    window.context = {
      ENV_hideParticipantIdentities: false,
    };
  });

  it('should render correctly when the participant is not the dominant speaker', () => {
    const wrapper = shallow(
      <Participant
        participant={{ identity: 'mockIdentity1' } as any}
        width="200px"
        height="100px"
        isDominantSpeaker={false}
        fontSize="56px"
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly when the participant is the dominant speaker', () => {
    const wrapper = shallow(
      <Participant
        participant={{ identity: 'mockIdentity1' } as any}
        width="300px"
        height="200px"
        isDominantSpeaker={true}
        fontSize="32px"
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly when the participant has disabled their video', () => {
    mockUseTracks.mockImplementationOnce(() => [
      { trackName: 'camera-123456', kind: 'video', trackSid: 'testVideoSid', isEnabled: false },
      { trackName: 'microphone-123456', kind: 'audio', trackSid: 'testAudioSid', isEnabled: true },
    ]);

    const wrapper = shallow(
      <Participant
        participant={{ identity: 'mockIdentity1' } as any}
        width="300px"
        height="200px"
        isDominantSpeaker={true}
        fontSize="32px"
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly when the participant has switched off their video', () => {
    mockUseTracks.mockImplementationOnce(() => [
      { trackName: 'camera-123456', kind: 'video', trackSid: 'testVideoSid', isEnabled: true, isSwitchedOff: true },
      { trackName: 'microphone-123456', kind: 'audio', trackSid: 'testAudioSid', isEnabled: true },
    ]);

    const wrapper = shallow(
      <Participant
        participant={{ identity: 'mockIdentity1' } as any}
        width="400px"
        height="600px"
        isDominantSpeaker={false}
        fontSize="56px"
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly when the participant has unpublished their video', () => {
    mockUseTracks.mockImplementationOnce(() => [
      { trackName: 'microphone-123456', kind: 'audio', trackSid: 'testAudioSid', isEnabled: true },
    ]);

    const wrapper = shallow(
      <Participant
        participant={{ identity: 'mockIdentity1' } as any}
        width="400px"
        height="600px"
        isDominantSpeaker={false}
        fontSize="56px"
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should display the muted icon when participant has disabled their audio', () => {
    mockUseTracks.mockImplementationOnce(() => [
      { trackName: 'camera-123456', kind: 'video', trackSid: 'testVideoSid', isEnabled: true },
      { trackName: 'microphone-123456', kind: 'audio', trackSid: 'testAudioSid', isEnabled: false },
    ]);
    const wrapper = shallow(
      <Participant
        participant={{ identity: 'mockIdentity1' } as any}
        width="300px"
        height="200px"
        isDominantSpeaker={true}
        fontSize="32px"
      />
    );

    expect(wrapper.find(MutedIcon).exists()).toBe(true);
  });

  it('should display the muted icon when participant has unpublished their audio', () => {
    mockUseTracks.mockImplementationOnce(() => [
      { trackName: 'camera-123456', kind: 'video', trackSid: 'testVideoSid', isEnabled: true },
    ]);
    const wrapper = shallow(
      <Participant
        participant={{ identity: 'mockIdentity1' } as any}
        width="400px"
        height="600px"
        isDominantSpeaker={true}
        fontSize="32px"
      />
    );

    expect(wrapper.find(MutedIcon).exists()).toBe(true);
  });

  describe('when ENV_hideParticipantIdentities is true', () => {
    beforeEach(() => {
      window.context = {
        ENV_hideParticipantIdentities: true,
      };
    });

    it('should not display the participants identity', () => {
      const wrapper = shallow(
        <Participant
          participant={{ identity: 'mockIdentity1' } as any}
          width="200px"
          height="100px"
          isDominantSpeaker={false}
          fontSize="56px"
        />
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('should not display the participants identity when their video is disabled', () => {
      mockUseTracks.mockImplementationOnce(() => [
        { trackName: 'camera-123456', kind: 'video', trackSid: 'testVideoSid', isEnabled: false },
        { trackName: 'microphone-123456', kind: 'audio', trackSid: 'testAudioSid', isEnabled: true },
      ]);

      const wrapper = shallow(
        <Participant
          participant={{ identity: 'mockIdentity1' } as any}
          width="300px"
          height="200px"
          isDominantSpeaker={true}
          fontSize="32px"
        />
      );

      expect(wrapper).toMatchSnapshot();
    });
  });
});
