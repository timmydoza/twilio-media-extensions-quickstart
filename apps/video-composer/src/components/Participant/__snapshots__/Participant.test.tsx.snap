// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Participant component should render correctly when the participant has disabled their video 1`] = `
<div
  className="participant"
  data-cy-participant="mockIdentity1"
  style={
    Object {
      "border": "solid 3px #7BEAA5",
      "height": "200px",
      "margin": 0,
      "width": "300px",
    }
  }
>
  <MediaTrack
    track={
      Object {
        "isEnabled": true,
        "kind": "audio",
        "trackName": "microphone-123456",
        "trackSid": "testAudioSid",
      }
    }
  />
  <div
    className="noVideo"
    style={
      Object {
        "fontSize": "32px",
      }
    }
  >
    mockIdentity1
  </div>
  <MediaTrack
    track={
      Object {
        "isEnabled": false,
        "kind": "video",
        "trackName": "camera-123456",
        "trackSid": "testVideoSid",
      }
    }
  />
  <p>
    mockIdentity1
  </p>
</div>
`;

exports[`the Participant component should render correctly when the participant has switched off their video 1`] = `
<div
  className="participant"
  data-cy-participant="mockIdentity1"
  style={
    Object {
      "border": "none",
      "height": "600px",
      "margin": 3,
      "width": "400px",
    }
  }
>
  <MediaTrack
    track={
      Object {
        "isEnabled": true,
        "kind": "audio",
        "trackName": "microphone-123456",
        "trackSid": "testAudioSid",
      }
    }
  />
  <div
    className="noVideo"
    style={
      Object {
        "fontSize": "56px",
      }
    }
  >
    mockIdentity1
  </div>
  <MediaTrack
    track={
      Object {
        "isEnabled": true,
        "isSwitchedOff": true,
        "kind": "video",
        "trackName": "camera-123456",
        "trackSid": "testVideoSid",
      }
    }
  />
  <p>
    mockIdentity1
  </p>
</div>
`;

exports[`the Participant component should render correctly when the participant has unpublished their video 1`] = `
<div
  className="participant"
  data-cy-participant="mockIdentity1"
  style={
    Object {
      "border": "none",
      "height": "600px",
      "margin": 3,
      "width": "400px",
    }
  }
>
  <MediaTrack
    track={
      Object {
        "isEnabled": true,
        "kind": "audio",
        "trackName": "microphone-123456",
        "trackSid": "testAudioSid",
      }
    }
  />
  <div
    className="noVideo"
    style={
      Object {
        "fontSize": "56px",
      }
    }
  >
    mockIdentity1
  </div>
  <p>
    mockIdentity1
  </p>
</div>
`;

exports[`the Participant component should render correctly when the participant is not the dominant speaker 1`] = `
<div
  className="participant"
  data-cy-participant="mockIdentity1"
  style={
    Object {
      "border": "none",
      "height": "100px",
      "margin": 3,
      "width": "200px",
    }
  }
>
  <MediaTrack
    track={
      Object {
        "isEnabled": true,
        "kind": "audio",
        "trackName": "microphone-123456",
        "trackSid": "testAudioSid",
      }
    }
  />
  <MediaTrack
    track={
      Object {
        "isEnabled": true,
        "kind": "video",
        "trackName": "camera-123456",
        "trackSid": "testVideoSid",
      }
    }
  />
  <p>
    mockIdentity1
  </p>
</div>
`;

exports[`the Participant component should render correctly when the participant is the dominant speaker 1`] = `
<div
  className="participant"
  data-cy-participant="mockIdentity1"
  style={
    Object {
      "border": "solid 3px #7BEAA5",
      "height": "200px",
      "margin": 0,
      "width": "300px",
    }
  }
>
  <MediaTrack
    track={
      Object {
        "isEnabled": true,
        "kind": "audio",
        "trackName": "microphone-123456",
        "trackSid": "testAudioSid",
      }
    }
  />
  <MediaTrack
    track={
      Object {
        "isEnabled": true,
        "kind": "video",
        "trackName": "camera-123456",
        "trackSid": "testVideoSid",
      }
    }
  />
  <p>
    mockIdentity1
  </p>
</div>
`;

exports[`the Participant component when ENV_hideParticipantIdentities is true should not display the participants identity 1`] = `
<div
  className="participant"
  data-cy-participant="mockIdentity1"
  style={
    Object {
      "border": "none",
      "height": "100px",
      "margin": 3,
      "width": "200px",
    }
  }
>
  <MediaTrack
    track={
      Object {
        "isEnabled": true,
        "kind": "audio",
        "trackName": "microphone-123456",
        "trackSid": "testAudioSid",
      }
    }
  />
  <MediaTrack
    track={
      Object {
        "isEnabled": true,
        "kind": "video",
        "trackName": "camera-123456",
        "trackSid": "testVideoSid",
      }
    }
  />
</div>
`;

exports[`the Participant component when ENV_hideParticipantIdentities is true should not display the participants identity when their video is disabled 1`] = `
<div
  className="participant"
  data-cy-participant="mockIdentity1"
  style={
    Object {
      "border": "solid 3px #7BEAA5",
      "height": "200px",
      "margin": 0,
      "width": "300px",
    }
  }
>
  <MediaTrack
    track={
      Object {
        "isEnabled": true,
        "kind": "audio",
        "trackName": "microphone-123456",
        "trackSid": "testAudioSid",
      }
    }
  />
  <div
    className="noVideo"
    style={
      Object {
        "fontSize": "32px",
      }
    }
  />
  <MediaTrack
    track={
      Object {
        "isEnabled": false,
        "kind": "video",
        "trackName": "camera-123456",
        "trackSid": "testVideoSid",
      }
    }
  />
</div>
`;
